### ISSUE TITLE
Analyze counterfactuals using microfounded demand model

### ISSUE THREAD

**Issue by tgerarden** [tgerarden/solar-tariffs-microfounded-demand#2]
**Goal:** Write code to compute counterfactual outcomes based on the microfounded demand estimates.

**Background:**
- For now we have decided to treat this demand specification as a robustness check, so I'm hoping I can hand this off to you to try to work out some of the details independently while I focused on writing up our main results.
- Unlike the constant elasticity demand specification, we don't have a simple analytical representation of quantity demanded as a function of price (because the residential demand elasticity comes from an aggregation across markets where we are simply aggregated the elasticity and not the entire demand function). To be frank, I have not thought hard about how that aggregation is done, but I've delegated it to Bryan and Daniel. Based on our conversations I think we'll need to analyze counterfactuals in terms of changes relative to baseline, and at least initially do it with a crude approximation to the true residential demand model.

Steps:
1. the first step is to make new cost predictions for firms due to changes in tariffs, manufacturing subsidies, manufacturing locations, etc. we have already done that for the constant elasticity demand model, and we can just use those
2. solve for equilibrium ΔQ and ΔP in response to a change in costs from 1 above
  - start with a candidate change in relative prices
  - use the firm FOC to make new predictions for market shares given the new cost predictions along with the candidate price change. censor the market shares at zero (no negative market shares). add the market shares up to get an aggregate market share prediction (e.g., an aggregate market share of 1.2 in a given time period would imply a counterfactual 20% larger market than the factual market)
  - use the demand model to compute the quantity change that would be consistent with the candidate price change:
    - **utility-scale demand**: for any price, we can compute the quantity demanded by the utility-scale sector since it's a static logit model. we would just need to compute the change in predicted market share for solar (vs other electricity generation techs), and then scale that by market size. i can't recall if we have all the inputs immediately available in the objects created by `estimate_model.R`. if not, see the code in `SolarTariff/regressions/utility_demand_estimation` for the necessary data. it should be pretty straightforward but let me know if you need more background on the model, etc. than i have already provided
    - **residential demand**: for simplicity we've decided to start with a rough approximation using the estimated elasticities. start with our vector of residential elasticities (note: the `SolarTariff/counterfactual/elasticity.xlsx` is semi-elasticities despite the name! i converted them to elasticities in the code, see line 488 of `estimate_model.R`). then multiply a hypothetical price change coming from the firm FOC with those elasticities (by time period)
    - combine the implied changes in Q from utility and residential (one for each time period)
  - compare the ΔQ from the firm FOC to the implied ΔQ from the demand models
    - if the change is firm FOC ΔQ is bigger than the demand ΔQ, lower the candidate price change. if the reverse, increase the candidate price change.
  - iterate until convergence of the ΔQ from firm FOC and demand
  - note that you will have to account for the market share of the competitive fringe using `ihs_usitc_ratio` (see existing code `/microfounded_supply_microfounded_demand/helper_functions/incentive_compatibility.R` for how we've done that for the constant elasticity model). our maintained assumption to date is that the market share of the competitive fringe is invariant to counterfactuals.
3. use the above to compute Δs in all the other counterfactual outcomes of interest (PS and CS, tariff revenue, environmental impacts, etc. etc.)
  - most of this will be identical to what we already have, just relying on the new ΔQs and ΔPs
  - we'll have to think about how to compute CS for this demand model, account for downstream demand subsidies, etc.
4. please type up a concise summary of the steps you end up having to use to do this, we might want to put it in the appendix

**Comment by tgerarden** [comment]
@hapham-econ i made some minor edits and may make more to the original post above, so please be sure to look at the version on github rather than in email

i'm going to commit your changes related to #1 now to compartmentalize tasks a bit in the version history

**Comment by tgerarden** [comment]
@hapham-econ i'm also about to post a few more minor tasks to #1 that i'd like you to finalize and then commit before you start on this

**Comment by hapham-econ** [comment]
@tgerarden Thank you for the detailed explanation. Here are my questions before starting this:

1. If I am understanding it correctly, the cost change is done at the start of the parts where counterfactuals are analyzed in `estimate_model.R`, and I can just use that to calculate the equilibrium, right? In fact, for each of these counterfactuals, I should include another version using the microfounded demand.

2. Most of the steps you have described are similar to what have been done in `incentive_compatibility.R`. My main job here is to modify the calculation of `w_hat` (price) and `Q_hat` (quantity) according to the steps you mentioned, right?

3. My plan is just to make another helper function called `incentive_compatibility_microfounded_demand.R` and include another part in the code where I run every counterfactual again using this function. Is that okay?

**Comment by hapham-econ** [comment]
@tgerarden I understand. I will work on this over the weekend. You should have an update on Monday!

**Comment by hapham-econ** [comment]
> utility-scale demand: for any price, we can compute the quantity demanded by the utility-scale sector since it's a static logit model. we would just need to compute the change in predicted market share for solar (vs other electricity generation techs), and then scale that by market size. i can't recall if we have all the inputs immediately available in the objects created by estimate_model.R. if not, see the code in SolarTariff/regressions/utility_demand_estimation for the necessary data. it should be pretty straightforward but let me know if you need more background on the model, etc. than i have already provided

I don't think the `merged_data` we used in `estimate_model.R` has information about utility-scale estimates (`delta` or `alpha` in the model)
![image](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/90290294/a1bf98ba-ae8c-4356-a3a0-352835ac611b)
so I looked at the code in the `SolarTariff/regressions/utility_demand_estimation`. I have located the parts in the code that I need. Can I modify the code so that I save these parts in the same file you outputted? I will need the intercept, coefficient and the total demand (later on you would need to run this file again on your machine). With this I can have the counterfactual quantity. 

> residential demand: for simplicity we've decided to start with a rough approximation using the estimated elasticities. start with our vector of residential elasticities (note: the SolarTariff/counterfactual/elasticity.xlsx is semi-elasticities despite the name! i converted them to elasticities in the code, see line 488 of estimate_model.R). then multiply the candidate price change with those elasticities (by time period) to get an implied quantity change. (this is a crude approximation) combine the implied changes in Q from utility and residential (one for each time period) 

You converted semielasticities to elasticities by multiplying by price (`w`), right? I understand how to get the change in quantity from this step. Once I got the change in quantity in utility and residential, I need to recalculate the `share_utility` and `share_comm_res` to calculate the weighted elasticity. How were these shares calculated? I tried to look at the `prep_cournot_data.R` file but I could not find how to. From my understanding, I need this weighted elasticities to plug into the firm FOC to get the change in firm quantity.
